### DISCUSSION TITLE
OAuth2 error: Invalid OAuth2 Redirect Parameters

### DISCUSSION THREAD

**Discussion by polypixeldev** [pocketbase/pocketbase#3474] [Q&A]
I'm currently trying to set up authentication for my app with React Native (Expo) on Android, using Pocketbase. I've enabled and set up OAuth2 providers for Google, Discord, and GitHub, and I'm able to login to the provider and authorize the app, but upon redirect to my Pocketbase instance (running on my Raspberry Pi, exposed using [Tailscale Funnel](https://tailscale.com/kb/1223/tailscale-funnel/), Pocketbase always responds with the same error:
```json
{
  "code": 400,
  "message": "Invalid OAuth2 redirect parameters.",
  "data": {}
}
```
I've configured the redirect URLs, client IDs, and client secrets for each provider, and looked at all the documentation and bug reports that I can find, but I do not know why it is not working. 

An example of the URL that is being launched (google, with hostname taken out): `https://accounts.google.com/o/oauth2/auth?client_id=69776731819-4o47ar7mvmv93i0g0kbfmc80om5lkjcb.apps.googleusercontent.com&code_challenge=vgnA1YDr9umYyfUiAF_lh0g6cWlZFR5bw4HB2733lOE&code_challenge_method=S256&response_type=code&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile%20https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email&redirect_uri=https%3A%2F%2F<PB_HOSTNAME>%2Fapi%2Foauth2-redirect`

An example of the URL that Google redirects me to (with hostname and code taken out): `https://<PB_HOSTNAME>/api/oauth2-redirect?code=<CODE>&scope=email+profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+openid&authuser=0&prompt=none`

**Comment by ganigeorgiev** [comment]
I'm not sure what could be the reason for the error.

Usually the _"Invalid OAuth2 redirect parameters"_ is thrown when the redirect is missing the `state` or `code` query parameters (in your redirect url you don't seem to have `state`).

How did you create and configure the Google OAuth2 app?

Could you share your auth code or create a minimal reproducible repo in [snack.expo.dev](https://snack.expo.dev/)?

**Comment by polypixeldev** [comment]
The Google OAuth2 app is configured as a web application, with just the one redirect URI set.
I initialize PocketBase with an AsyncAuthStore, as shown here
```
const store = new AsyncAuthStore({
  save: async (serialized) => AsyncStorage.setItem("pb_auth", serialized),
});
const pb = new Pocketbase(process.env.EXPO_PUBLIC_POCKETBASE_URL, store);
```

Upon clicking a button, the following code runs:
```
const userData = await pb
            .collection("users")
            .authWithOAuth2({
              provider: "google",
              urlCallback: async (url) => {
                Linking.openURL(url);
              },
            })
            .catch((err) => {
              console.log({ ...err });
            });

          console.log(userData);
```
I just tried adding a state query parameter to the URL in the urlCallback before opening the URL (set to "test"), and if I do that, upon redirect I get a different error with code 404 and message "Missing or invalid OAuth2 subscription client".
I'm not sure why the state parameter is missing initially, but it seems that's the issue.

**Reply by ganigeorgiev** [reply]
_Just a side-note: in the above `AsyncAuthStore` instance you are not loading the `initial` value, meaning that on app restart `pb.authStore` will not be populated with the previously stored auth state. If you use v0.18.2 of the SDK you can initialize it as shown in https://github.com/pocketbase/js-sdk#asyncauthstore._

**Comment by polypixeldev** [comment]
I found the issue - I had been using a different eventsource implementation for react native, but when I switched back to `react-native-sse` it worked. Thanks for your help!
